### DISCUSSION TITLE
Create log drain from vercel cli or web gui

### DISCUSSION THREAD

**Discussion by tianhuil** [vercel/vercel#7542] [Ideas] [answered]
Currently, the only way to create a custom log drain is to follow the rather tedious instructions provided by this [third-party open-source tool](https://github.com/valeriangalliat/vercel-custom-log-drain).  This requires spinning up a server to handle an OAuth2 flow just to provide a webhook URL to Vercel (which likely points to another server that I am spinning up to actually handle the logs).  The current architecture is designed for marketplace integrations where a third party is involved.

However, if I just want to write a simple log custom logdrain, I do not need OAuth2 or a separate server just to submit a URL to Vercel.  I would rather either:
1. Use the cli to enter a URL callback, e.g.:
`vercel log-drain create https://example.com/callback deployment deployment-ready`
2. Use the web GUI to do the same thing (actually preferred).

Can we at the very least add the cli functionality (#1) to this repo?

**Comment by davidstenstroem** [comment]
I have also stumbled upon [this resource](https://www.codejam.info/2021/09/vercel-custom-log-drain.html) on how to make log drains.

But I agree 100% that it would be very nice to have what you suggest.

**Comment by zoul** [comment]
We would love to see log drain setup from the CLI, too. We would even donate some engineer time to help implementing the feature if that helped.

In the meantime we have started writing [an integration that lets you setup any HTTP log drain](https://github.com/cesko-digital/vercel-http-log-drain). It’s better than the one linked in the parent post in that it lets you sign in using the OAuth token later and manage the drains without reinstalling the integration all the time. And we have included a sample HTTP drain to store the logs in Elastic. The code is rough at the moment, but we already use it in production.

(That said, it’s still a hack and a CLI solution would be better.)

**Reply by frosty5689** [reply]
Any instructions on how to get the client id/client secret for Vercel OAuth? Is it generated when you create a custom integration? The form to create an integration is super convuluted asking for product images of 3:2 aspect ratio but keeps failing validation despite using correct aspect ratio images.

**Comment by lumenwrites** [comment]
Any updates on this?

I'd really love to be able to send my logs to a custom url, but it seems surprisingly difficult to figure out.

**Comment by lumenwrites** [comment]
@zoul Hey, thank you so much for sharing this! I'm trying to use this integration, but I'm getting an error:
> The value `undefined` is not of type `string`, but is of type `undefined` when trying to decode the key `clientId` in `{"error":{"code":"bad_request","message":"Only integrations can subscribe to log-drains"}}`

<img width="1436" alt="Screen Shot 2022-09-08 at 14 08 04" src="https://user-images.githubusercontent.com/4761235/189129920-eb73a4fc-7991-4a7f-b8c9-54821ad04c72.png">

What am I doing wrong? Can you please help me understand where can I take the `clientId`, and where should I put it in order to make this work?

Edit:
Oh, it looks like `clientId` isn't the issue. It shows up in an error because `decodeLogDrain` can't find it, because `createLogDrain` returns an error:
> {"code":"bad_request","message":"Only integrations can subscribe to log-drains"}

Do you know why that could happen?

**Comment by and-poulsen** [comment]
+1
This is insanely tedious

**Comment by zoul** [comment, marked as answer]
This is now possible with the new [Configurable Log Drains](https://vercel.com/changelog/configurable-log-drains).
